mailbox/store/src/main/java/org/apache/james/mailbox/store/PreDeletionHooks.java

```diff
@@ -32,20 +35,27 @@
 import reactor.core.scheduler.Schedulers;
 
 public class PreDeletionHooks {
-    public static final PreDeletionHooks NO_PRE_DELETION_HOOK = new PreDeletionHooks(ImmutableSet.of());
+    public static final PreDeletionHooks NO_PRE_DELETION_HOOK = new PreDeletionHooks(ImmutableSet.of(), new NoopMetricFactory());
+
+    static final String PRE_DELETION_HOOK_METRIC_NAME = "preDeletionHook";
 
     private final Set<PreDeletionHook> hooks;
+    private final MetricFactory metricFactory;
 
     @Inject
-    public PreDeletionHooks(Set<PreDeletionHook> hooks) {
+    public PreDeletionHooks(Set<PreDeletionHook> hooks, MetricFactory metricFactory) {
         this.hooks = hooks;
+        this.metricFactory = metricFactory;
     }
 
     public Mono<Void> runHooks(PreDeletionHook.DeleteOperation deleteOperation) {
         return Flux.fromIterable(hooks)
             .publishOn(Schedulers.elastic())
             .limitRate(1)
-            .flatMap(hook -> hook.notifyDelete(deleteOperation))
+            .flatMap(hook -> metricFactory.runPublishingTimerMetric(PRE_DELETION_HOOK_METRIC_NAME,
+                MDCBuilder.create()
+                    .addContext(MDCBuilder.ACTION, "PRE_DELETION_HOOK")
+                    .wrapArround(() -> hook.notifyDelete(deleteOperation))))
```

Comment:
Your MDC is useless as `hook.notifyDelete(deleteOperation)` is a publisher and might very well be executed by anouther thread.

Reply:
After discussion it has been decided that it might be not worth it to use MDC